libqtile/backend/x11/core.py

```diff
@@ -232,16 +234,23 @@ def on_config_load(self, initial) -> None:
         """Assign windows to groups"""
         assert self.qtile is not None
 
+        # Ensure that properties are initialised at startup
+        self.update_client_lists()
+
         if not initial:
             # We are just reloading config
             for win in self.qtile.windows_map.values():
                 if type(win) is window.Window:
                     win.set_group()
             return
 
+        hook.subscribe.client_focus(self.check_stacking)
+        hook.subscribe.client_managed(self._on_client_managed)
+        hook.subscribe.client_killed(self._on_client_killed)
```

Comment:
(Happy to iterate on this in a future PR, ideally adding a TODO here for the time being). As the backends listen for events from clients and hardware/X server, they are ultimately responsible for firing hooks, so it feels unnatural to subscribe to hooks in the backend too. E.g. currently the X backend subscribes to none, only fires them. The wayland backend subscribes only to `startup_complete` though for the same reason I think that should actually use an asyncio `call_soon` handler instead.

The idea is similar to a producer-consumer relationship, where the backend emits events and other python code within the qtile process listens for them as a consumer, and ideally we'd keep that relationship clean.

Concrete suggestions:

- drop 247, instead do `self.core.check_stacking()` here: https://github.com/qtile/qtile/blob/master/libqtile/backend/x11/window.py#L1044
- drop 248 and `Core._on_client_managed`, instead do `self.core.update_client_lists(); self.change_layer()` here: https://github.com/qtile/qtile/blob/master/libqtile/backend/x11/window.py#L1462
- drop 249 and `Core._on_client_killed`, instead do `self.update_client_lists()` after `unmanage()` in these two places:
  - https://github.com/qtile/qtile/blob/master/libqtile/backend/x11/core.py#L728
  - https://github.com/qtile/qtile/blob/master/libqtile/backend/x11/core.py#L754

Reply:
I think we should do this now. However, the `client_managed` bit you've linked to is in the `Static` class. I think the hook is fired for other windows by the manager here: https://github.com/qtile/qtile/blob/4ecf1bdef46d68da56145a654701ed355dcad64b/libqtile/core/manager.py#L677

Reply:
`unmanage` is more complicated as we need the window that's passed when the hook is fired. We could have `self.qtile.unmanage` return the window so it can be used by the calling method.